File: random/ValidParentheses_20.java
package leetcode;

import java.util.HashMap;
import java.util.Stack;

/**
 * Valid Parentheses
 * 有效的括号
 */
public class ValidParentheses_20 {
    private static HashMap<Character, Character> mappings;

    public static void main(String[] args) {
        String test ="{[()]}";
        initMap();
        System.out.println(isValid(test));
    }

    private static void initMap() {
        mappings = new HashMap<Character, Character>();
        mappings.put(')', '(');
        mappings.put('}', '{');
        mappings.put(']', '[');
    }

    public static boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (mappings.containsKey(c)) {
              char topElement = stack.empty()?'#':stack.pop();
              if (topElement!=mappings.get(c)){
                  return false;
              }
            } else {
                stack.push(c);
            }
        }

        return stack.isEmpty();
    }
}

File: random/kt/ValidParentheses_20.kt
package leetcode.random.kt
import java.util.*
import kotlin.collections.HashMap

/**
 * 20. 有效括号
给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。
有效字符串需满足：
左括号必须用相同类型的右括号闭合。
左括号必须以正确的顺序闭合。


示例 1：

输入：s = "()"
输出：true
示例 2：

输入：s = "()[]{}"
输出：true
示例 3：

输入：s = "(]"
输出：false
示例 4：

输入：s = "([)]"
输出：false
示例 5：

输入：s = "{[]}"
输出：true


来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/valid-parentheses
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */


/**
 *
 */
object ValidParentheses_20 {
    @JvmStatic
    fun main(args: Array<String>) {
        val parentheses = "()[]{}"
        println( validParentheses(parentheses))
    }

    private fun validParentheses(parentheses: String): Boolean {
        val map = HashMap<Char, Char>()
        val stack = Stack<Char>()
        map[')'] = '('
        map[']'] = '['
        map['}'] = '{'

        for (i in parentheses.indices) {
            val value = parentheses[i]
            if (map.containsKey(value)) {
                val element = if (stack.empty()) '#' else stack.pop()
                if (element!= map[value]) return false
            } else {
                stack.push(value)
            }
        }
        return stack.isEmpty()
    }

}